dashboards-observability/public/components/visualizations/charts/data_table/data_table.tsx

```diff
@@ -3,102 +3,487 @@
+      resizable: true,
+      filter: true,
+      flex: 1,
+      minWidth: 100,
+      suppressMenu: true,
+    };
+  }, []);
 
-  // Sorting
-  const [sortingColumns, setSortingColumns] = useState([]);
-  const onSort = useCallback(
-    (sortingColumns) => {
-      setSortingColumns(sortingColumns);
+  const onPageSizeChanged = useCallback(
+    (val) => {
+      setPageSize(val);
+      gridRef.current.api.paginationSetPageSize(val);
+      setActivePage(0);
+      gridRef.current.api.paginationGoToPage(0);
+      if (isFullScreen) {
+        gridRefFullScreen.current.api.paginationSetPageSize(val);
+        gridRefFullScreen.current.api.paginationGoToPage(0);
+      }
     },
-    [setSortingColumns]
+    [isFullScreen]
   );
 
-  // Sort data
-  let data = useMemo(() => {
-    return [...raw_data].sort((a, b) => {
-      for (let i = 0; i < sortingColumns.length; i++) {
-        const column = sortingColumns[i];
-        const aValue = a[column.id];
-        const bValue = b[column.id];
+  const paginationNumberFormatter = useCallback((params) => {
+    return '[' + params.value.toLocaleString() + ']';
+  }, []);
 
-        if (aValue < bValue) return column.direction === 'asc' ? -1 : 1;
-        if (aValue > bValue) return column.direction === 'asc' ? 1 : -1;
+  const selectDensityHandler = useCallback((value) => {
+    setSelectedRowDensity({ ...value });
+    gridRef.current.api.forEachNode((rowNode) => {
+      if (rowNode.data) {
+        rowNode.setRowHeight(value.height);
       }
-
-      return 0;
     });
-  }, [sortingColumns]);
+    gridRef.current.api.onRowHeightChanged();
+  }, []);
 
-  // Pagination
-  data = useMemo(() => {
-    const rowStart = pagination.pageIndex * pagination.pageSize;
-    const rowEnd = Math.min(rowStart + pagination.pageSize, data.length);
-    return data.slice(rowStart, rowEnd);
-  }, [data, pagination]);
+  const columnVisiblityHandler = useCallback((visible: boolean, feild: any) => {
+    const isExisting = columnVisibility.findIndex((i) => i === feild);
+    if (visible) {
+      if (isExisting > -1) {
+        columnVisibility.splice(isExisting, 1);
+        gridRef?.current?.columnApi?.setColumnsVisible([feild], true);
+      }
+    } else {
+      columnVisibility.push(feild);
+      gridRef?.current?.columnApi?.setColumnsVisible([feild], false);
+    }
+    setColumnVisibility([...columnVisibility]);
+  }, []);
 
-  const columns = fields.map((field) => {
-    return {
-      id: field.name,
+  // pagination
+  const pageCount = Math.ceil(raw_data.length / pageSize);
+  const [activePage, setActivePage] = useState(0);
+  const goToPage = ({ selected }: { selected: number }) => {
+    setActivePage(selected);
+    gridRef.current.api.paginationGoToPage(selected);
+    if (isFullScreen) {
+      gridRefFullScreen.current.api.paginationGoToPage(selected);
+    }
+  };
+
+  const setIsFullScreenHandler = (val: boolean) => {
+    setIsFullScreen(val);
+  };
+
+  useEffect(() => {
```

Comment:
All the hooks are conventionally defined at the top of the file.

Reply:
moved at top